fix: readonly is not stripped from private prop

bartlomieju opened the pull request:
Closes #915

bartlomieju committed:
fix: readonly is not stripped from private prop

bartlomieju commented on ecmascript/codegen/src/lib.rs:
```diff
@@ -901,10 +901,10 @@ impl<'a> Emitter<'a> {
-            keyword!("readonly");
-            space!();
-        }
+        // if n.readonly {
+        //     keyword!("readonly");
+        //     space!();
+        // }
```
This is probably not right, @kdy1 could you give me a short explanation on how emitter works? `readonly` should be emitted if we emit TypeScript, but should be stripped when emitting JavaScript. I can't find any place where that's decided.

kdy1 replied:
@bartlomieju It's determined by the data of node.

If you want special behavior only for typescript, you can change swc_emca_transforms::typescript::strip. In this case, you can modify `fold_private_prop` to remove readonly flag.

kdy1 replied:
@bartlomieju If you are ok with it, I'll add some commits to the pr to fix the bug.

bartlomieju replied:
@kdy1 się, however it seems more bugs were found: 
- It's not just private props: target/debug/deno cache --unstable --no-check -r std/mime/multipart.ts. (The source line is even more haywire here, it's caused by readonly left on public MultipartReader fields.)

- I think there are way more bugs than with readonly.
- target/debug/deno cache --unstable --no-check -r std/path/win32.ts: generates with messed up backslash escape in template literal.
- target/debug/deno cache --unstable --no-check -r std/node/module.ts: generates with messed up hexadecimal escape sequence.

- - target/debug/deno cache --unstable --no-check -r std/path/_globrex.ts: backslash problem described earlier.
- target/debug/deno cache --unstable --no-check -r std/wasi/snapshot_preview1.ts: same.
- target/debug/deno cache --unstable --no-check -r std/ws/test.ts: doesn't strip enum.
That's all the errors I could yield in std, some might be blocked by others.

I'll add tests for all of them in a couple hours, and your commits are more than welcome

kdy1 replied:
I fixed private property issue and there are some known - although no issue exists on github - escape issues with codegen.
An enum not being stripped is bit surprising.

Test codes are in the deno repository, is it correct?

kdy1 replied:
@bartlomieju  There was a bug in the parser and codegen. I send a [pr](https://github.com/bartlomieju/swc/pull/1)

bartlomieju replied:
Thanks for explanation @kdy1, I'll add more tests to codegen for the escape sequences.

kdy1 committed:
Transcompile enums declared in functions

kdy1 committed:
Private property is not readonly after compilation

kdy1 committed:
Add tests

kdy1 committed:
Bump versions

kdy1 committed:
Fix tests

kdy1 committed:
Revert "fix: readonly is not stripped from private prop"

This reverts commit e714a6ba80bc79c287462e14578c85edfe8d32c0.

kdy1 committed:
Fix escape problem

kdy1 committed:
Fix escape problem of template literal

bartlomieju committed:
Merge pull request #1 from kdy1/issue-915

Fixes

bartlomieju committed:
remove print statements

kdy1 commented on ecmascript/codegen/src/lib.rs:
```diff
@@ -2308,15 +2308,10 @@ fn unescape(s: &str) -> String {
     let mut result = String::with_capacity(s.len() * 6 / 5);
     let mut chars = s.chars();
 
-    println!("S: {:?}", s);
```
Oh... Nice catch.
I completely forgot about it as tests passed and nothing was shown.

bartlomieju committed:
strip readonly from class prop

bartlomieju commented:
@kdy1 I found a few more problems:
```ts
// Input:
export class MultipartReader {
  readonly newLine = encoder.encode("\r\n");
  ... snip ...
}

// Output:
export class MultipartReader {
  readonly newLine = encoder.encode("\r\n"); <--- doesn't remove readonly
  ... snip ...
}
```

**Already fixed the one above**

```ts
// Input:
relResolveCacheIdentifier = `${parent.path}\x00${request}`;

// Output:
relResolveCacheIdentifier = `${parent.path}\x0${request}`;  // <--- \x00 transformed into \x0
```

```ts
// Input:
const data = text.encode(`${arg}\0`);

// Output:
const data = text.encode(`${arg}\`); // <--- \0 transformed into \
```

```ts
// Input:
export class MultipartReader {
  readonly newLine = encoder.encode("\r\n");
  readonly newLineDashBoundary = encoder.encode(`\r\n--${this.boundary}`);
  readonly dashBoundaryDash = encoder.encode(`--${this.boundary}--`);
 ... snip ...
}

// Output:
export class MultipartReader {
    newLine = encoder.encode("\r\n");
    newLineDashBoundary = encoder.encode(`\
\
--${this.boundary}`);   // <--- completely messed up \r\n replaced with actual newline token
    dashBoundaryDash = encoder.encode(`--${this.boundary}--`);
  ... snip ...
}
```

kdy1 commented:
I'll take a look

bartlomieju commented:
It's strange cause you added exact same tests for the escape problems 😕

kdy1 committed:
Ensure that swc can lex / parse

export class MultipartReader {
  readonly newLine = encoder.encode("\r\n");
}

kdy1 committed:
Fix mistake

kdy1 committed:
Add tests about `\0` becoming `\`

kdy1 committed:
Remove useless tests

swc_ecma_transfroms check if ast matches and identical test cannot be done with it.

kdy1 committed:
Add tests for #915

kdy1 committed:
Fix printing of hex

kdy1 committed:
Fix printing of hex and some escape characters

kdy1 committed:
Patch for #450

kdy1 commented:
@bartlomieju I think I fixed it. I'm not sure if it is correct fix, though.

bartlomieju commented:
@kdy1 I just pulled the code and tested again, those two things are still failing:
```ts
const data = text.encode(`${arg}\0`);
// strips to:
const data = text.encode(`${arg}\`);
```

```ts
export class MultipartReader {
  readonly newLine = encoder.encode("\r\n");
  readonly newLineDashBoundary = encoder.encode(`\r\n--${this.boundary}`);
  readonly dashBoundaryDash = encoder.encode(`--${this.boundary}--`);
 ... snip ...
}

// strips to
export class MultipartReader {
    newLine = encoder.encode("\r\n");
    newLineDashBoundary = encoder.encode(`\
\
--${this.boundary}`); 
```

kdy1 commented:
I'll try using raw files instead of string literals in test. I think the issue is related to it.

bartlomieju commented:
@kdy1 this test fails:
```
#[test]
fn tpl_escape_5() {
    test_from_to(
        r#"const data = text.encode(`${arg}\0`);"#,
        r#"const data = text.encode(`${arg}\0`);"#,
    );
}
```

```
---- tests::tpl_escape_5 stdout ----
--------------------
Source:
const data = text.encode(`${arg}\0`);
Pos: BytePos(0) ~ BytePos(37)

thread 'tests::tpl_escape_5' panicked at 'assertion failed: `(left == right)`
  left: `const data = text.encode(`${arg}\`);`,
 right: `const data = text.encode(`${arg}\0`);`', ecmascript/codegen/src/tests.rs:94:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::tpl_escape_5
```

bartlomieju committed:
add failing tests

bartlomieju committed:
fix one of the tests

kdy1 committed:
Fix \0

kdy1 commented:
@bartlomieju I fixed \0 problem

bartlomieju commented:
> @bartlomieju I fixed \0 problem

Thank you, I confirmed that problematic files are now properly stripped. How you seen my change to new line handling? https://github.com/swc-project/swc/pull/916/commits/a9bd9446f273e12f99dd8533e2606ec9248e249d

kdy1 commented:
I think it is great. That's what I should do after removing `s = s.replace("\\\\", "\\")`

bartlomieju commented:
Awesome 👍  we'll be releasing new version of Deno tomorrow so I expect more bugs will be uncovered by users 😅  hopefully we'll be able to fix them as we go
